README.md
# AWS Batch Architecture for AlphaFold 
-----
## Overview
Proteins are large biomolecules that play an important role in the body. Knowing the physical structure of proteins is key to understanding their function. However, it can be difficult and expensive to determine the structure of many proteins experimentally. One alternative is to predict these structures using machine learning algorithms. Several high-profile research teams have released such algorithms, including [AlphaFold 2](https://deepmind.com/blog/article/alphafold-a-solution-to-a-50-year-old-grand-challenge-in-biology), [RoseTTAFold](https://www.ipd.uw.edu/2021/07/rosettafold-accurate-protein-structure-prediction-accessible-to-all/), and others. Their work was important enough for Science magazine to name it the ["2021 Breakthrough of the Year"](https://www.science.org/content/article/breakthrough-2021).

Both AlphaFold 2 and RoseTTAFold use a multi-track transformer architecture trained on known protein templates to predict the structure of unknown peptide sequences. These predictions are heavily GPU-dependent and take anywhere from minutes to days to complete. The input features for these predictions include multiple sequence alignment (MSA) data. MSA algorithms are CPU-dependent and can themselves require several hours of processing time.

Running both the MSA and structure prediction steps in the same computing environment can be cost inefficient, because the expensive GPU resources required for the prediction sit unused while the MSA step runs. Instead, using a high-performance computing (HPC) service like [AWS Batch](https://aws.amazon.com/batch/) allows us to run each step as a containerized job with the best fit of CPU, memory, and GPU resources.

This repository includes the CloudFormation template, Jupyter Notebook, and supporting code to run the Alphafold v2 algorithm on AWS Batch. 

-----
## Architecture Diagram
![AWS Batch Architecture for AlphaFold](imgs/aws-alphafold-arch.png)

-----
## First time setup
### Deploy the infrastructure stack

1. Choose **Launch Stack**:

    [![Launch Stack](imgs/LaunchStack.jpg)](https://console.aws.amazon.com/cloudformation/home#/stacks/create/review?templateURL=https://aws-batch-architecture-for-alphafold-public-artifacts.s3.amazonaws.com/main/alphafold-cfn-packaged.yaml)

2. Specify the following required parameters:
  - For **Stack Name**, enter a value unique to your account and region.
  - For **StackAvailabilityZone** choose an availability zone.
3. Provide a name for your stack, leave the other parameters as their default, and select **Next**.
4. Select **I acknowledge that AWS CloudFormation might create IAM resources with custom names**.
5. Choose **Create stack**.
6. Wait 15 minutes for AWS CloudFormation to create the infrastructure stack and AWS CodeBuild to build and publish the AWS-AlphaFold container to Amazon Elastic Container Registry (Amazon ECR).

### Launch SageMaker Notebook
(If **LaunchSageMakerNotebook** set to Y)
1. Navigate to [SageMaker](https://console.aws.amazon.com/sagemaker)
2. Select **Notebook** > **Notebook instances**.
3. Select the **AWS-Alphafold-Notebook** instance and then **Actions** > **Open Jupyter** or **Open JupyterLab**.

![Sagemaker Notebook Instances](imgs/notebook-nav.png)

### Clone Notebook Repository
(If **LaunchSageMakerNotebook** set to N)
1. Navigate to [CodeCommit](https://console.aws.amazon.com/codesuite/codecommit).
2. Select the aws-alphafold repository that was just created and copy the clone URL.
3. Use the URL to clone the repository into your Jupyter notebook environment of choice, such as SageMaker Studio.

### Populate FSx for Lustre File System
1. If you set the **DownloadFsxData** parameter to **Y**, CloudFormation will automatically start a series of Batch jobs to populate the FSx for Lustre instance with a number of common sequence databases. If you set this parameter to **N** you will instead need to manually populate the file system. Once the CloudFormation stack is in a CREATE_COMPLETE status, you can begin populating the FSx for Lustre file system with the necessary sequence databases. To do this automatically, open a terminal in your notebooks environment and run the following commands from the AWS-AlphaFold directory:

```
> pip install -r notebooks/notebook-requirements.txt
> python notebooks/download_ref_data.py
```

2. It will take 3 to 4 hours to populate the file system, depending on your region. You can track its progress by navigating to the file system in the FSx for Lustre console.

### Cleaning Up
1. To delete all provisioned resources from from your account, navigate to [Cloud Formation](https://console.aws.amazon.com/cloudformation), select your stack, and then **Delete**. 

-----
## Optional Template Parameters
  - Select a different value for **AlphaFoldVersion** if you want to include another version of the public Alphafold repo in your Batch job container.
  - Specify a different value for **CodeRepoBucket** and **CodeRepoKey** if you want to populate your AWS-Alphafold CodeCommit repository with custom code stored in another S3 bucket.
  - Select a different value for **FSXForLustreStorageCapacity** if you want to provision a larger file system. The default value of 1200 GB is sufficient to store compressed instances of the Alphafold parameters and all standard sequence databases, BFD (small and reduced), Mgnify, PDB70, PDB mmCIF, Uniclust30, Uniref90, UniProt, and PDB SeqRes datasets. However, you can specify a larger value if you need to store additional data.
  - Select a different value for for **FSxForLustreThroughput** if you have unique performance needs. The default is 500 MB/s/TB. Select a higher value for performance-sensitive workloads and a lower value for cost-sensitive workloads.
  - Select "Y" for **LaunchSageMakerNotebook** if you want to launch a managed sagemaker notebook instance to quickly run the provided Jupyter notebook.
  - Provide values for the **VPC**, **Subnet**, and **DefaultSecurityGroup** parameters to use existing network resources. If one or more of those parameters are left empty, CloudFormation will create a new VPC and FSx for Lustre instance for the stack.
  - Provide values for the **FileSystemId** and **FileSystemMountName** parameters to use an existing FSx for Lustre file system. If one or more of these parameters are left empty, CloudFormation will create a new file system for the stack.
  - Select "Y" for **DownloadFsxData** to automatically populate the FSx for Lustre file system with common sequence databases.

-----
## Usage
Use the provided `AWS-AlphaFold.ipynb` notebook to submit sequences for analysis and download the results.

-----
## Performance
To determine the optimal compute settings, we used the [CASP14 target list](https://predictioncenter.org/casp14/targetlist.cgi) to test various CPU, memory, and GPU settings for the data preparation and prediction jobs. Using the full BFD database can increase the duration of data preparation jobs by as much as 10x. However the resulting increase in MSA coverage can increase the maximum pLDDT scores for some targets.

![Job Performance Depends on Target Size and DB Type](imgs/performance.png)
![Using Reduced BFD Can Affect Prediction Quality](imgs/plddt.png)

Based on this analysis, we recommend the following AWS Batch compute resource settings:

### Data Preparation (Reduced BFD)
- CPsU: 4 vCPU
- Memory: 16 GiB
- GPUs: 0

### Data Preparation (Full BFD)
- CPUs: 16 vCPU
- Memory: 32 GiB
- GPUs: 0

### Prediction (Sequence Length < 700)
- CPUs: 4
- Memory: 16 GiB
- GPUs: 1

### Prediction (Sequence Length > 700)
- CPUs: 16
- Memory: 64 GiB
- GPUs: 1

-----
## Cost Estimation
Follow these steps to estimate the per-run costs asociated with a protein of size X:

1. Pick either the full or reduced bfd database
1. Find the length of your target sequence on the x-axis.
1. Use the plotted curves to identify the estimated data prep and prediction job durations.
1. Refer to the [EC2 On-Demand pricing page](https://aws.amazon.com/ec2/pricing/on-demand/) to obtain the hourly rate for the data prep job instance type equivalent (m5.xlarge or c5.4xlarge, depending on bfd database type) and prediction job instance type equivalent (g4dn.xlarge or g4dn.4xlarge, depending on sequence length).

![AWS Batch Run Time Estimation](imgs/cost-estimation.png)

For example, analyzing a 625-residue sequence using the reduced bfd database will take approximately 0.3 hours of data prep time, plus 1 hour of prediction time. As of February 2022, the on-demand rate for a m5.xlarge instance in the us-east-1 Region is $0.192/hr. and the rate for a g4dn.xlarge instance is $0.526/hr., for a total estimated cost of $0.72 per run. Please note that this pricing is subject to change at any time.

-----
## Security

See [CONTRIBUTING](CONTRIBUTING.md#security-issue-notifications) for more information.

-----
## License

This project is licensed under the Apache-2.0 License.

-----
## Additional Information

### AlphaFold Repository
Please visit https://github.com/deepmind/alphafold for more information about the AlphaFold algorithm.

### Citations
The original AlphaFold 2 paper is
```
@Article{AlphaFold2021,
  author  = {Jumper, John and Evans, Richard and Pritzel, Alexander and Green, Tim and Figurnov, Michael and Ronneberger, Olaf and Tunyasuvunakool, Kathryn and Bates, Russ and {\v{Z}}{\'\i}dek, Augustin and Potapenko, Anna and Bridgland, Alex and Meyer, Clemens and Kohl, Simon A A and Ballard, Andrew J and Cowie, Andrew and Romera-Paredes, Bernardino and Nikolov, Stanislav and Jain, Rishub and Adler, Jonas and Back, Trevor and Petersen, Stig and Reiman, David and Clancy, Ellen and Zielinski, Michal and Steinegger, Martin and Pacholska, Michalina and Berghammer, Tamas and Bodenstein, Sebastian and Silver, David and Vinyals, Oriol and Senior, Andrew W and Kavukcuoglu, Koray and Kohli, Pushmeet and Hassabis, Demis},
  journal = {Nature},
  title   = {Highly accurate protein structure prediction with {AlphaFold}},
  year    = {2021},
  volume  = {596},
  number  = {7873},
  pages   = {583--589},
  doi     = {10.1038/s41586-021-03819-2}
}
```
The AlphaFold-Multimer paper is 
```
@article {AlphaFold-Multimer2021,
  author       = {Evans, Richard and O{\textquoteright}Neill, Michael and Pritzel, Alexander and Antropova, Natasha and Senior, Andrew and Green, Tim and {\v{Z}}{\'\i}dek, Augustin and Bates, Russ and Blackwell, Sam and Yim, Jason and Ronneberger, Olaf and Bodenstein, Sebastian and Zielinski, Michal and Bridgland, Alex and Potapenko, Anna and Cowie, Andrew and Tunyasuvunakool, Kathryn and Jain, Rishub and Clancy, Ellen and Kohli, Pushmeet and Jumper, John and Hassabis, Demis},
  journal      = {bioRxiv}
  title        = {Protein complex prediction with AlphaFold-Multimer},
  year         = {2021},
  elocation-id = {2021.10.04.463034},
  doi          = {10.1101/2021.10.04.463034},
  URL          = {https://www.biorxiv.org/content/early/2021/10/04/2021.10.04.463034},
  eprint       = {https://www.biorxiv.org/content/early/2021/10/04/2021.10.04.463034.full.pdf},
}
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.4] - 2022-06-24

### Changed
- Added support for AlphaFold v2.2.2


## [1.0.3] - 2022-06-03

### Fixed
- Resolved issue where timings.json file was not properly passed between the data prep and folding jobs (Identified by PR https://github.com/aws-samples/aws-batch-architecture-for-alphafold/pull/3)
- Resolved compatibility issue with Protobuf version (https://github.com/deepmind/alphafold/issues/478)

### Changed

- Added support for AlphaFold v2.2.0
- Added support for existing VPCs and FSx for Lustre instances
- Added option to automatically download reference data to FSx
- Decreased data prep job cost by as much as 75% by using spot instance types.
- Updated data download script to pull sequence databases and parameters from S3
- Decreased data download time by 41% by retrieving ref data in parallel
- Refactored CloudFormation template to use nested stacks
- Added CloudFormation tests with Taskcat
- Updated folding container to Python 3.9, Ubuntu 20.04, and mamba

## [1.0.2] - 2022-05-20

### Fixed

### Changed

- Added additional download scripts to help address the instability of the public PDB mmcif mirror.
- Moved download jobs into private subnet for increased security.

## [1.0.1] - 2022-02-25

### Fixed

- Fix bug in download-Dockerfile that was causing download jobs to run longer than needed.
- Fix typo in section headers in AWS-AlphaFold notebook.

### Changed

- Add table of contents to AWS-AlphaFold notebook.

## [1.0] - 2022-02-24

### Changed

- Open Source Release

## [0.2] - 2022-02-23

### Changed

- Documentation updates
- Security and IP updates

## [0.1] - 2022-02-10

### Added

- Initial Release
- Submit 1- and 2-step protein-folding jobs to AWS Batch using a Jupyter Notebook.


cdkv1/lokafold-app/app.py
import os
import sys
import logging
import boto3
from chalice import Chalice
from chalicelib import (
    validate_input,
    create_job_name,
    upload_fasta_to_s3,
    submit_batch_alphafold_job,
    get_batch_job_info
)
app = Chalice(app_name="lokafold-app")

boto_session = boto3.session.Session()
region = boto_session.region_name
batch = boto_session.client("batch", region_name=region)

S3_BUCKET = os.environ["FASTA_BUCKET"]

logging.basicConfig(
    format="%(asctime)s %(levelname)-s:%(name)s: %(message)s",
    level=logging.INFO,
    datefmt="%H:%M:%S",
    stream=sys.stderr
)
logger = logging.getLogger(__name__)

@app.route("/", methods=["POST", "GET"])
def index():
    request = app.current_request
    body = request.json_body
    if request.method == 'GET':
        job_id = request.query_params["job_id"]
        # job_id_2 = request.query_params["prediction_job_id"]
        status = get_batch_job_info(job_id)
        # status_2 = get_batch_job_info(job_id_2)
        response = {
            "job": status,
        }
        return response
    elif request.method == 'POST':
        input_sequences = list(body["sequences"].values())
        input_ids = list(body["sequences"].keys())
        db_preset = body["db_preset"]
        # cpu = body["cpu"]
        # memory = body["memory"]
        # gpu = body["gpu"]
        # run_features_only = body["run_features_only"]
        
        # Validate input for invalid aminoacid residues
        input_sequences, model_preset = validate_input(input_sequences)
        sequence_length = len(max(input_sequences))
        
        if db_preset == "reduced_dbs":
            prep_cpu = 4
            prep_mem = 16
            prep_gpu = 0

        else:
            prep_cpu = 16
            prep_mem = 32
            prep_gpu = 0

        if sequence_length < 700:
            predict_cpu = 4
            predict_mem = 16
            predict_gpu = 1
        else:
            predict_cpu = 16
            predict_mem = 64
            predict_gpu = 1
        

        # Upload file to s3 bucket
        job_name = create_job_name()
        object_key = upload_fasta_to_s3(
            input_sequences,
            input_ids,
            S3_BUCKET,
            job_name
        )
        # Submit jobs to batch
        try:
            step_1_response = submit_batch_alphafold_job(
                job_name=str(job_name),
                fasta_paths=object_key,
                output_dir=job_name,
                db_preset=db_preset,
                model_preset=model_preset,
                s3_bucket=S3_BUCKET,
                cpu=prep_cpu,
                memory=prep_mem,
                gpu=prep_gpu,
                run_features_only=True,
            )
            
            step_2_response = submit_batch_alphafold_job(
                job_name=str(job_name),
                fasta_paths=object_key,
                output_dir=job_name,
                db_preset=db_preset,
                model_preset=model_preset,
                s3_bucket=S3_BUCKET,
                cpu=predict_cpu,
                memory=predict_mem,
                gpu=predict_gpu,
                features_paths=os.path.join(job_name, job_name, "features.pkl"),
                depends_on=step_1_response["jobId"],
            )
        except Exception as err:
            raise f"Error submiting the jobs {err}"
        
        response = {
            "feature_extraction_job_response": step_1_response,
            "prediction_job_response": step_2_response,
        }
        return response
